lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java

```diff
@@ -231,9 +285,20 @@ public Identifier getNetworkConnectionServiceId() {
 
   @Override
   public void close() throws Exception {
-    LOG.log(Level.FINE, "Shutting down");
-    this.nameServiceRegisteringStage.close();
-    this.nameServiceUnregisteringStage.close();
-    this.transport.close();
+    if (isClosed.compareAndSet(false , true)) {
+      LOG.log(Level.FINE, "Shutting down");
+      this.nameServiceRegisteringStage.close();
+      this.nameServiceUnregisteringStage.close();
+      this.transport.close();
+      unregisterAllConnectionFactories();
+    } else {
+      throw new NetworkRuntimeException("Multiple requests to close the NetworkConnectionService");
```

Comment:
In my experience, calling close() multiple times is tolerated. This may be a bit harsh.